mlflow/types/schema.py

```diff
@@ -406,3 +458,280 @@ def __eq__(self, other) -> bool:
+            raise TypeError(
+                "Expected mlflow.models.signature.Datatype or str for the 'type' "
+                f"argument, but got {self.type.__class__}"
+            )
+        if self.type == DataType.binary:
+            raise MlflowException(
+                f"Binary type is not supported for parameters, ParamSpec '{self.name}'"
+                "has type 'binary'",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        # This line makes sure repr(self) works fine
+        self._default = default
+        self._default = self.validate_type_and_shape(repr(self), default, self.type, self.shape)
+
+    @classmethod
+    def validate_param_spec(
+        cls, value: Union[DataType, List[DataType], None], param_spec: "ParamSpec"
+    ):
+        return cls.validate_type_and_shape(
+            repr(param_spec), value, param_spec.type, param_spec.shape
+        )
+
+    @classmethod
+    def enforce_param_datatype(cls, name, value, t: DataType):
+        """
+        Enforce the value matches the data type.
+
+        The following type conversions are allowed:
+
+        1. int -> long, float, double
+        2. long -> float, double
+        3. float -> double
+        4. any -> datetime (try conversion)
+
+        Any other type mismatch will raise error.
+
+        :param name: parameter name
+        :param value: parameter value
+        :param t: expected data type
+        """
+        if value is None:
+            return value
+
+        if not np.isscalar(value):
+            raise MlflowException(
+                f"Value should be a scalar for param {name}, got {value}", INVALID_PARAMETER_VALUE
+            )
+
+        if getattr(DataType, f"is_{t.name}")(value):
+            return value
+
+        if (
+            (DataType.is_integer(value) and t in (DataType.long, DataType.float, DataType.double))
+            or (DataType.is_long(value) and t in (DataType.float, DataType.double))
+            or (DataType.is_float(value) and t == DataType.double)
+            or t == DataType.datetime
+        ):
+            try:
+                if t == DataType.datetime:
+                    return np.datetime64(value)
+                return np.array(value, dtype=t.to_numpy()).take(0)
+            except ValueError as e:
+                raise MlflowException(
+                    f"Failed to convert value {value} from type {type(value).__name__} "
+                    f"to {t} for param {name}",
+                    INVALID_PARAMETER_VALUE,
+                ) from e
+
+        raise MlflowException(
+            f"Incompatible types for param {name}. Can not safely convert {type(value).__name__} "
+            f"to {t}.",
+            INVALID_PARAMETER_VALUE,
+        )
+
+    @classmethod
+    def validate_type_and_shape(
+        cls,
+        spec: str,
+        value: Union[DataType, List[DataType], None],
+        value_type: DataType,
+        shape: Optional[Tuple[int, ...]],
```

Comment:
do we need to use Tuple here or would List suffice? I'm wary of non-JSON-serializable types if ever the shape spec would need to be encoded

Reply:
I think tuple works fine
<img width="250" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/5f34fc4a-b484-41ca-b520-f7045dbd515d">

Reply:
It converts to a list :) 

``` python
import json

# Original tuple
original_tuple = (1, 2, 3)

# Serialize the tuple to a JSON string
serialized = json.dumps(original_tuple)

print(f'Serialized: {serialized}')

# Deserialize the JSON string back to a Python object
deserialized = json.loads(serialized)

print(f'Deserialized: {deserialized}')

# Check if the deserialized object equals the original tuple
print(f'Equality check: {original_tuple == deserialized}')
```

```sh
Serialized: [1, 2, 3]
Deserialized: [1, 2, 3]
Equality check: False
```

Reply:
Yes, but we convert it back to tuple when constructing the shape
https://github.com/mlflow/mlflow/blob/578d326f5af7ecfed412c2ee802e363d0d374c56/mlflow/types/schema.py#L477

Or you prefer we convert this to a list instead? I'm totally fine though :D